### PR TITLE
[ONNX] Introduce FX-ONNX dispatcher

### LINKED ISSUE #100542: [ONNX] Implement a dispatcher for the converter
All torch_lib functions from onnxscript now has a `op_schema` field that has the signatures of the functions. We can use this information to dispatch.

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102048
* __->__ #100660

Needs https://github.com/microsoft/onnxscript/pull/721

The current FX exporter is using manually maintained dictionary to map ATen op to its OnnxFunction. However, the issue arises when ATen op has overloads or OnnxFunction has overloads, which is not resolvable by the one to one mapping . For example, `aten::arange` has onverloads: `aten::arange.start` and `aten::arange.start_step`, or for `aten::argmax`, torchlib provides two function: aten_argmax, and aten_argmax_dim.

This PR utilizes newly introduced [ONNX OpSchema](https://github.com/microsoft/onnxscript/pull/626) to match the input arguments of an ATen operator to find the correct overload.

### OnnxRegistry

Heavily reference on [TorchScript Registry](https://github.com/pytorch/pytorch/pull/84382). The only difference is that in FX registry, an ATen operator with specific opset version is mapped to a list of overloaded functions.

* No longer use global registry. The registry is initialized in `ResolvedExportOptions` with torchlib, and will be exposed to users in the future.
* Multiple opset version layer is kept through `_SymbolicFunctionGroup` , but torchlib now only supports 18.
* Basic API of custom operator support: `register`, `unregister`, and `is_register_op` are kept for future development. To further complete them, the follow-up PRs should address:
    - How to allow users to remove/override specific overload? Using OpSchema to differentiate?
    - User registers a new overload with the same OpSchema as one of registered overload. 
 
### OnnxDispatcher

Dispatch ATen operators to the matched overload by comparing OpSchema with input arguments.

* `OpSchemaWrapper` wrap the onnx schema, and record matching score.
* `dispatch` uses `OpSchemaWrapper` to compare data types to find the best matched overload. If the match isn't perfect, record warning in diagnostics.
* `dispatch_opset_version` is referenced from #84382 and kept, but torchlib doesn't support opset version != 18.
* Because right now (1) OnnxFunction arguments are manually typed, and (2) ORT could unfollow ONNX type spec, we relax the schema match with `matching score system`.
* To include more supports:  the follow-up PRs should address:
    - How to add op.Cast with autocast? In torchlib or converter?
    - The need of type promotion can be captured by dispatcher, but needs OpSchema shows the T1/T2 information.

#### The matching score system:
This is a temporary solution to how we target the correct ONNX overloads given that we only have manually annotated arguments (potentially inaccurate schema) and limited supports on AttributeProto.

1. Perfect match exam: If all arguments/kwargs are all matched, return the function without any warnings.
2. Best match exam: The system add the each correct matching input counts orderly, and subtract the symmetrical difference between their attributes to calculate the matching score. And select the one with the highest score in the end. If the selection is not a perfect match, a warning message is sent to SARIF.

#### Example of overloads

1. Different types: Caused by the difference between the ONNX spec and PyTorch.

The matching system finds the correct one.

```python
@torch_op("aten::mul")
def aten_mul(self: TReal, other: TReal) -> TReal:
    ...

@torch_op("aten::mul")
def aten_mul_bool(self: BOOL, other: BOOL) -> BOOL:
    ...
```

2. Optional dim: caused by unsupported op.OptionalHasElement (will support on opset version == 20). dim could be "None"

```python
@torch_op("aten::argmax", trace_only=True)
def aten_argmax(
    self: TrealOrUInt8, dim: Optional[int] = None, keepdim: bool = False
) -> TrealOrUInt8:
    ...

@torch_op("aten::argmax", private=True)
def _aten_argmax_dim(self: TrealOrUInt8, dim: int, keepdim: bool = False) -> TrealOrUInt8:
    ...
```

This case is impossible to differentiate, as they both might have dim in kwargs, so in this case, please make sure you turn the one with `dim: int` to private function.

3. Optional dtype: dtype could be "unprovided". The difference from 2 is that dtype would not be None.

```python
@torch_op("aten::new_full")
def aten_new_full(self: TTensor, size: INT64, fill_value: TTensor) -> TTensor:
    ...

@torch_op("aten::new_full")
def aten_new_full_dtype(self: TTensor, size: INT64, fill_value: TTensor, dtype: int) -> TTensor:
    ...
```

Depends on dtype is provided or not, matching system will dispatch the ATen op to the correct one.

4. `None` and `[]` and `NoneType` are considered failing the match. 

5. Two functions have the same score is recorded into SARIFs. 

### TODOs

1. Type promotion can be captured by dispatcher only if OpSchema can provide it. However, the implementation of "graph-level" pass vs "in-op"" promotion can be further discussed in https://github.com/microsoft/onnxscript/issues/563.
5. torchlib should provide the "opset version" to OnnxRegistry.
7. How to expose OnnxRegistry with custom add/remove ops APIs nneds to be further discussed.

Co-authored-by: Justin Chu <justinchuby@microsoft.com>

### CODE REVIEW

**Comment by justinchuby** [comment]
Looks like need to implement get in Registry. Also need to bump onnxscript commit hash

**Comment by titaiwangms** [comment]
> Looks like need to implement get in Registry. Also need to bump onnxscript commit hash

Yup, I thought rushing a draft for @BowenBao to get a general idea of how this is going, so we can have an opportunity to discuss before you gone for vacation.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/function_dispatcher.py:152]: Does `Registry` support cloning?

I'm thinking in the context of when we move to non global dispatcher & registry for `dynamo_export`. Also consider custom op registering where registry is modified.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/function_dispatcher.py:145]: not sure how 'f' is reflected this condition checking :) Should each overload has an op schema?

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/function_dispatcher.py:64]: Who is responsible for creating the cast nodes?

What is the design contract of this casting logic? What is the relationship of this with torch's "implicit type promotion", and "[implicit integral promotion](https://github.com/pytorch/pytorch/pull/100265#discussion_r1183025474)"?

Comment (fatcat-z) [reply torch/onnx/_internal/fx/function_dispatcher.py:64]: IMO, ONNX-Script torch_lib should provide type-matched function(overload) for each aten/prims op, so we should not have such case (type-mismatched).

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (fatcat-z) [inline torch/onnx/_internal/fx/function_dispatcher.py:87]: onnx_args? onnx_kwargs?

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (justinchuby) [reply torch/onnx/_internal/fx/function_dispatcher.py:64]: > Who is responsible for creating the cast nodes?
> 
> What is the design contract of this casting logic? What is the relationship of this with torch's "implicit type promotion", and "[implicit integral promotion](https://github.com/pytorch/pytorch/pull/100265#discussion_r1183025474)"?

The dispatcher will select the function. A separate type promotion logic will create the node.

> IMO, ONNX-Script torch_lib should provide type-matched function(overload) for each aten/prims op, so we should not have such case (type-mismatched).

We will when the onnx op does not support the input at all (asin etc.)

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (justinchuby) [reply torch/onnx/_internal/fx/function_dispatcher.py:145]: Right. Should have been

function_opschema = OpSchemaWrapper(f)

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:64]: This one can be put on hold. I think it's a "nice to have" feature.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:152]: Let me reference on the old one, and see how we can support converter side registration.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:145]: Oops!

**Comment by thiagocrepaldi** [comment]
I will look into this and compare with https://microsoft-my.sharepoint.com/:w:/p/thiagofc/EW-5Q3jWhFNMtQHHtPpJiAQB-P2qAcVRkYjfbmeSddnjWA?e=UTzumt exporter design.

We need to make sure type promotion, variants, etc can be handled without too many changes in future

**Comment by titaiwangms** [comment]
> I will look into this and compare with https://microsoft-my.sharepoint.com/:w:/p/thiagofc/EW-5Q3jWhFNMtQHHtPpJiAQB-P2qAcVRkYjfbmeSddnjWA?e=UTzumt exporter design.
> 
> We need to make sure type promotion, variants, etc can be handled without too many changes in future
> 
> Also for reference: [microsoft/onnxscript#563](https://github.com/microsoft/onnxscript/issues/563)

Thanks! I will check Registry and Dispatcher API design session.

**Review by thiagocrepaldi — COMMENTED** [review]
Is there any design doc available?

Topics to discuss:

* type promotion mechanism
  * implicit and explicit?
* multiple opset version
  * different OSes have different opset support
* autocast support
  * explicit casts, each operator has its own rules on when to down or upcast
* custom operator support
* not necessarily needed initially, but consider how a possible fallback mechanism could be introduced without major hacking

File: torch/onnx/_internal/fx/registration.py
Comment (titaiwangms) [inline torch/onnx/_internal/fx/registration.py:198]: Figure out way to provide opset version from torchlib

File: torch/onnx/_internal/fx/registration.py
Comment (titaiwangms) [inline torch/onnx/_internal/fx/registration.py:62]: Use set

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [inline torch/onnx/_internal/fx/function_dispatcher.py:60]: If there is not perfect match, we choose the most matched one.

File: torch/onnx/_internal/fx/registration.py
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/registration.py:177]: This dictionary might be too simple to efficiently support multiple opset versions, multiple op overloads in aten and in onnx, custom ops and maybe differentiating prims from aten in the future (not sure whether we will ever support prims, but if there is a chance, maybe we should cover this now)

At https://github.com/pytorch/pytorch/pull/100660/files#diff-ccff3d48c69a764390d32023acc4ba29de4b9cb8b596bd926cb0c6d481990d0cR65 there is a todo saying "# FIXME(titaiwang): How to remove a specific function instead of whole overloads?" which feels is directly associated on how we store the symbolic functions

File: torch/onnx/_internal/fx/decomposition_table.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/decomposition_table.py:19]: This was an old code, and I simply annotated it. I think it's some aten ops are not strictly under torch._ops.OpOverload. I can try finding those ops.

File: torch/onnx/_internal/fx/decomposition_table.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/decomposition_table.py:75]: I incline to not touch the logics here (this was here), as it's not in this PR scope. I know it would be great if we can really make multiple opsets refactor easier, if this PR can do this and do that. But all I really want to focus in this PR is to dispatch the overloads I mentioned in the description.

File: torch/onnx/_internal/fx/decomposition_table.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/decomposition_table.py:62]: Not sure, this is not added in this PR.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:95]: I hide onnxscript under TYPE_CHECKING for this file. Any suggestion?

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:94]: I will try finish this one, but as you know, args and kwargs from torch is kind of freestyle.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:109]: prim is not supported yet, we can do this when it's supported.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:192]: Disagree, this doesn't return overloads

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:275]: hmmm I kind of hope that we can utilize OpSchema as other pass consultant. But you are right, currently it should go private for now.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:320]: No, T is not implemented yet. perfect match means the all inputs types fall into the allowed constraints.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:279]: Great suggestion thanks!

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/function_dispatcher.py:320]: Fair enough, just include that info on docstring and it should be fine

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/function_dispatcher.py:60]: Having this captured by SARIF is fine, but it should not replace docstring documentation. docstring help us compare the actual behavior from the expected behavior. SARIF only helps with the former

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/function_dispatcher.py:95]: I meant that `) -> Union[onnxscript.OnnxFunction, onnxscript.TracedOnnxFunction]:` should work, without having to surround it with `""`. does it not?

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/function_dispatcher.py:94]: `io_adapter.py` might have this annotation

File: torch/onnx/_internal/fx/decomposition_table.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/decomposition_table.py:62]: I thought it was added at https://github.com/pytorch/pytorch/pull/100660/files#diff-df6cf326cf4c77dd4c6feaf75f8357a5ad5101bfe209c5df700d8fd5a085884eR87-R97

File: torch/onnx/_internal/fx/registration.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/registration.py:177]: TBH, I don't have much thoughts on registry design, because I think we will definitely have a huge refactor on registry when we decide how to proceed with multiple opset version. On the other hand, to complete dispatcher, I need a draft of registry which leads us to current situation. 

I am not against refactor the registry in this PR, as the dispatcher does engage a lot with it. It's just with people gone for vacation, I don't see more plans on this, and where should we head to with "the design that can make future work easier". We can discuss in the team meeting about this. I am fine even if we need to hold the dispatcher until a great plan on registry is made.

File: torch/onnx/_internal/fx/decomposition_table.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/decomposition_table.py:62]: No, it's not. These are not originated from this PR. 

If you are asking `_create_onnx_supports_op_overload_table()`. This is a function to find overloadpacket with their overloads, not ONNX.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:95]: I got error saying NameError about onnxscript. It seems TYPE_CHECKING allowing it to be checked in static, but I have to quote it.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:94]: Not really the same, this guy is from `_wrap_fx_args_as_onnxscript_args` which we haven't really annotated it well.

Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:94]: I just found I annotated in `perfect_match`, but they are actually using the same args lol

File: torch/onnx/_internal/fx/registration.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/registration.py]: Done

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py]: Updated in OpSchemaWrapper descriptions using the examples from this PR description.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py]: Done

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py]: I thought about your suggestion. I think even if we finally decide to do type promotion in dispatcher, I image it would be after we find an overload, which means that function should be called next to dispatch(), not inside it.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:254]: Yes, the group refers to the `_SymbolicFunctionGroup`. You can find the same usage in torchscript exporter. It's the same structure, and same API. My intention is to keep the registry as the way we used to, so when we have an idea of how do proceed multiple opset versions, and custom ops support, we would be able to modify on the registry we are familiar with.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py]: Done

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:332]: I added some descriptions. LMK if they are more clear.

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:44]: Make sense

Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py]: Done

**Comment by titaiwangms** [comment]
I can't repro the error in local. It seems we need ONNX submodule version bumps ...

**Review by thiagocrepaldi — APPROVED** [review]
Overall LGTM, just make sure CI is green before merging it

**Pull request closed without merging**
